### DISCUSSION TITLE
Add manifest(s) to PHAR distribution

### DISCUSSION THREAD

**Discussion by llaville** [composer/composer#11429] [Show and tell]
Hello,

**Add manifest(s) to PHAR distribution** : Goal is to know contents of dependencies embedded with a PHAR distribution !

> **Edit** 
> While I'm debugging an auto-update manifests feature on my project https://github.com/llaville/box-manifest for upcoming version 3.5.0, I've discovered a nasty issue with Composer.
>
> See related discussion about the SF conflict on new topic https://github.com/composer/composer/discussions/11430

To learn more about PHAR contents, I suggest to add at least one manifest (choose your format).

I propose to make it with my project https://github.com/llaville/box-manifest ( it's a [BOX-project](https://github.com/box-project/box) compatible with manifests support) 

**box.json.dist**
```json
{
    "output": "bin/composer-php8.phar",
    "directories": ["bin", "res", "src", "vendor"],
    "files-bin": ["LICENSE"],
    "check-requirements": false,
    "compression": "GZ",
    "stub": "stub.php",
    "files-bin": [
        "console.txt",
        "manifest.txt",
        "sbom.json",
        "sbom.xml"
    ],
    "map": [
        { "console.txt": ".box.manifests/console.txt" },
        { "manifest.txt": ".box.manifests/manifest.txt" },
        { "sbom.xml": ".box.manifests/sbom.xml" },
        { "sbom.json": ".box.manifests/sbom.json" }
    ]
}
```  

Builds the PHAR with following command:  `box-manifest-3.4.phar box:compile --composer-bin=bin/composer -c box.json.dist`

```text
 // PHAR: 1026 files (4.72MB)
 // You can inspect the generated PHAR with the "info" command.

 // Memory usage: 54.62MB (peak: 55.11MB), time: 2secs

? Setting manifests
  - console.txt > text/plain :
  - manifest.txt > text/plain :
  - sbom.json > application/json : CycloneDX v1.4
  - sbom.xml > text/xml : CycloneDX v1.4
```

And if you want to test to see all manifests included (for demo purpose only): one is enough (choose your format)

- `bin/composer-php8.phar  --manifest sbom.xml`
- `bin/composer-php8.phar  --manifest sbom.json`
- `bin/composer-php8.phar  --manifest manifest.txt`
or simply 
- `bin/composer-php8.phar  --manifest`  (first found by priority in the `stub.php` file) 
  
![composer-255_php-8_manifests](https://user-images.githubusercontent.com/364342/232080381-5114159e-ac97-49fe-acd7-b7fe8c6d619a.png)
![composer-255_php-8_manifests_2](https://user-images.githubusercontent.com/364342/232080423-d7074e3c-0dc8-4f10-85bd-f1675a8f22f3.png)

**Comment by llaville** [comment]
I forget to give a result given with my `post-update-cmd` that work fine when I run it with my new Composer 2.5.5 PHP 8 (Symfony Console v6) PHAR distrib.  The auto-update manifests work as expected !

![box-manifest-350](https://user-images.githubusercontent.com/364342/232082576-608e5932-62e6-4a47-ae8c-1a3283508860.png)

**Comment by stof** [comment]
It looks like this discussion tries to cover 2 totally different feature requests:

- exposing a manifest of what is included in the PHAR (as said in the title)
- shipping a separate PHAR using newer dependencies for PHP 8

I suggest you to separate the 2 topics, otherwise the discussion will be a mess.

**Comment by llaville** [comment]
@stof Done !

**Comment by llaville** [comment]
Related to my comment https://github.com/composer/composer/discussions/11429#discussioncomment-5617196, here are the full implementation that allow these results (take care also to read the https://github.com/composer/composer/discussions/11430 topic).

- https://github.com/llaville/box-manifest/commit/81b11ffd867be19c7c35b628a565c1bfb8a4c844
- https://github.com/llaville/box-manifest/commit/c37603ab8395b287e0e367fdc561a89bacaab2eb

**Comment by Seldaek** [comment]
Generally speaking yes I think adding a CycloneDX SBOM would make sense. I'd definitely prefer having a tiny piece of code in Composer that handles outputting that file and then generating the SBOM with some third party tool at build time when we create the phar.

I'd rather not add a dependency on box nor runtime dependencies for Composer for this. If you have pointers for how to achieve this feel free to open an issue with them, but tbh this isn't super high on my priority list right now.

**Comment by llaville** [comment]
Even if I'm always on side to adopt standards, I don't think SBOM is readable for Human (whatever format used : XML or JSON). Computer is the best to proceed to check results by automation, but humans are not able to have a good summary with SBOM like `composer show` do !
So, as I won't let you lose your time and mine, I'll close this discussion.

I'm pretty sure that @jkowalleck is the best candidate to propose a solution with the https://github.com/CycloneDX/cyclonedx-php-composer project !
